core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java

```diff
@@ -98,4 +111,80 @@ public T getFirst(Object value) {
     Set<T> all = mIndexMap.get(value);
     return all == null ? null : Iterables.getFirst(all, null);
   }
+
+  @Override
+  public Iterator<T> iterator() {
+    return new NonUniqueFieldIndexIterator();
+  }
+
+  @Override
+  public int size() {
+    int totalSize = 0;
+    for (ConcurrentHashSet<T> innerSet : mIndexMap.values()) {
+      totalSize += innerSet.size();
+    }
+    return totalSize;
+  }
+
+  /**
+   * Specialized iterator for {@link NonUniqueFieldIndex}.
+   *
+   * This is needed to support consistent removal from the set and the indices.
+   */
+  private class NonUniqueFieldIndexIterator implements Iterator<T> {
+    /**
+     * Iterator of {@link NonUniqueFieldIndex#mIndexMap}. This Iterator keeps track of the
+     * inner set which is under iteration.
+     */
+    private final Iterator<ConcurrentHashSet<T>> mIndexMapIterator;
+    /**
+     * Iterator inside each inner set. This iterator keeps track of the objects.
+     */
+    private Iterator<T> mIndexIterator;
```

Comment:
maybe better names could be `mObjectIterator` for this one, and `mIndexIterator` for the one above.

Or, if the above is still `mIndexMapIterator` this should be `mObjectSetIterator`.